File: animation/src/main/java/me/wyne/wutils/animation/AnimationStep.java
package me.wyne.wutils.animation;

import org.jetbrains.annotations.NotNull;

/**
 * A single stage of an {@link Animation}.
 *
 * <p>The first step is started by the owning {@link Animation}; every step thereafter is
 * started by its predecessor. A step drives its {@link AnimationRunnable} on a schedule and
 * then hands off to the next queued step (see the concrete implementations in this package
 * for how, and on which thread, that happens).</p>
 *
 * <p>{@link #close()} and {@link #_finalize()} both default to a no-op; implementations only
 * need to override the one relevant to how they hold resources.</p>
 */
@FunctionalInterface
public interface AnimationStep extends AutoCloseable, Finalizable {
    /**
     * Starts this step as part of the given animation.
     */
    void run(@NotNull Animation animation);
    @Override
    default void close() {}
    @Override
    default void _finalize() {}
}

File: animation/src/main/java/me/wyne/wutils/animation/AnimationStep.kt
package me.wyne.wutils.animation

interface AnimationStep : AutoCloseable {
    fun run(animation: Animation)
}

abstract class BaseAnimationStep(
    protected val runnable: AnimationRunnable,
    protected val delay: Long = 0,
    protected val period: Long = 0,
    protected val duration: Long = 0
) : AnimationStep {
    protected var ticksElapsed: Long = 0

    override fun run(animation: Animation) {
        ticksElapsed = 0
        createTask(animation)
    }

    override fun close() {
        runnable.close()
    }

    private fun createTask(animation: Animation) {
        if (period == 0L)
            runOnce(animation)
        else
            runRepeating(animation)
    }

    protected abstract fun runOnce(animation: Animation)
    protected abstract fun runRepeating(animation: Animation)
}